File: mRemoteV1/UI/UI.Window.ScreenshotManager.vb
Imports System.ComponentModel
Imports System.Drawing.Imaging
Imports System.IO
Imports mRemote3G.App
Imports mRemote3G.Forms
Imports mRemote3G.Messages
Imports mRemote3G.My.Resources
Imports mRemote3G.Tools
Imports WeifenLuo.WinFormsUI.Docking

Namespace UI

    Namespace Window
        Public Class ScreenshotManager
            Inherits Base

#Region "Form Init"

            Friend WithEvents msMain As MenuStrip
            Friend WithEvents mMenFile As ToolStripMenuItem
            Friend WithEvents mMenFileSaveAll As ToolStripMenuItem
            Friend WithEvents mMenFileRemoveAll As ToolStripMenuItem
            Friend WithEvents cMenScreenshot As ContextMenuStrip
            Private components As IContainer
            Friend WithEvents cMenScreenshotCopy As ToolStripMenuItem
            Friend WithEvents cMenScreenshotSave As ToolStripMenuItem
            Friend WithEvents dlgSaveSingleImage As SaveFileDialog
            Friend WithEvents dlgSaveAllImages As FolderBrowserDialog
            Friend WithEvents flpScreenshots As FlowLayoutPanel

            Private Sub InitializeComponent()
                Me.components = New System.ComponentModel.Container
                Me.flpScreenshots = New FlowLayoutPanel
                Me.msMain = New MenuStrip
                Me.mMenFile = New ToolStripMenuItem
                Me.mMenFileSaveAll = New ToolStripMenuItem
                Me.mMenFileRemoveAll = New ToolStripMenuItem
                Me.cMenScreenshot = New ContextMenuStrip(Me.components)
                Me.cMenScreenshotCopy = New ToolStripMenuItem
                Me.cMenScreenshotSave = New ToolStripMenuItem
                Me.dlgSaveSingleImage = New SaveFileDialog
                Me.dlgSaveAllImages = New FolderBrowserDialog
                Me.msMain.SuspendLayout()
                Me.cMenScreenshot.SuspendLayout()
                Me.SuspendLayout()
                '
                'flpScreenshots
                '
                Me.flpScreenshots.Anchor = CType((((AnchorStyles.Top Or AnchorStyles.Bottom) _
                                                   Or AnchorStyles.Left) _
                                                  Or AnchorStyles.Right),
                                                 AnchorStyles)
                Me.flpScreenshots.AutoScroll = True
                Me.flpScreenshots.Location = New Point(0, 26)
                Me.flpScreenshots.Name = "flpScreenshots"
                Me.flpScreenshots.Size = New Size(542, 296)
                Me.flpScreenshots.TabIndex = 0
                '
                'msMain
                '
                Me.msMain.Font = New Font("Microsoft Sans Serif", 8.25!, FontStyle.Regular, GraphicsUnit.Point,
                                          CType(0, Byte))
                Me.msMain.Items.AddRange(New ToolStripItem() {Me.mMenFile})
                Me.msMain.Location = New Point(0, 0)
                Me.msMain.Name = "msMain"
                Me.msMain.RenderMode = ToolStripRenderMode.Professional
                Me.msMain.Size = New Size(542, 24)
                Me.msMain.TabIndex = 1
                Me.msMain.Text = "MenuStrip1"
                '
                'mMenFile
                '
                Me.mMenFile.DropDownItems.AddRange(New ToolStripItem() {Me.mMenFileSaveAll, Me.mMenFileRemoveAll})
                Me.mMenFile.Image = Resources.File
                Me.mMenFile.Name = "mMenFile"
                Me.mMenFile.Size = New Size(51, 20)
                Me.mMenFile.Text = "&File"
                '
                'mMenFileSaveAll
                '
                Me.mMenFileSaveAll.Image = Screenshot_Save
                Me.mMenFileSaveAll.Name = "mMenFileSaveAll"
                Me.mMenFileSaveAll.Size = New Size(128, 22)
                Me.mMenFileSaveAll.Text = "Save All"
                '
                'mMenFileRemoveAll
                '
                Me.mMenFileRemoveAll.Image = Screenshot_Delete
                Me.mMenFileRemoveAll.Name = "mMenFileRemoveAll"
                Me.mMenFileRemoveAll.Size = New Size(128, 22)
                Me.mMenFileRemoveAll.Text = "Remove All"
                '
                'cMenScreenshot
                '
                Me.cMenScreenshot.Items.AddRange(New ToolStripItem() {Me.cMenScreenshotCopy, Me.cMenScreenshotSave})
                Me.cMenScreenshot.Name = "cMenScreenshot"
                Me.cMenScreenshot.Size = New Size(100, 48)
                '
                'cMenScreenshotCopy
                '
                Me.cMenScreenshotCopy.Image = Screenshot_Copy
                Me.cMenScreenshotCopy.Name = "cMenScreenshotCopy"
                Me.cMenScreenshotCopy.Size = New Size(99, 22)
                Me.cMenScreenshotCopy.Text = "Copy"
                '
                'cMenScreenshotSave
                '
                Me.cMenScreenshotSave.Image = Screenshot_Save
                Me.cMenScreenshotSave.Name = "cMenScreenshotSave"
                Me.cMenScreenshotSave.Size = New Size(99, 22)
                Me.cMenScreenshotSave.Text = "Save"
                '
                'dlgSaveSingleImage
                '
                Me.dlgSaveSingleImage.Filter =
                    "Graphics Interchange Format File (.gif)|*.gif|Joint Photographic Experts Group Fi" &
                    "le (.jpeg)|*.jpeg|Joint Photographic Experts Group File (.jpg)|*.jpg|Portable Ne" &
                    "twork Graphics File (.png)|*.png"
                Me.dlgSaveSingleImage.FilterIndex = 4
                '
                'ScreenshotManager
                '
                Me.ClientSize = New Size(542, 323)
                Me.Controls.Add(Me.flpScreenshots)
                Me.Controls.Add(Me.msMain)
                Me.HideOnClose = True
                Me.Icon = Screenshot_Icon
                Me.MainMenuStrip = Me.msMain
                Me.Name = "ScreenshotManager"
                Me.TabText = "Screenshots"
                Me.Text = "Screenshots"
                Me.msMain.ResumeLayout(False)
                Me.msMain.PerformLayout()
                Me.cMenScreenshot.ResumeLayout(False)
                Me.ResumeLayout(False)
                Me.PerformLayout()
            End Sub

#End Region

#Region "Form Stuff"

            Private Sub ScreenshotManager_Load(sender As Object, e As EventArgs) Handles Me.Load
                ApplyLanguage()
            End Sub

            Private Sub ApplyLanguage()
                mMenFile.Text = Language.Language.strMenuFile
                mMenFileSaveAll.Text = Language.Language.strSaveAll
                mMenFileRemoveAll.Text = Language.Language.strRemoveAll
                cMenScreenshotCopy.Text = Language.Language.strMenuCopy
                cMenScreenshotSave.Text = Language.Language.strSave
                dlgSaveSingleImage.Filter = Language.Language.strSaveImageFilter
                TabText = Language.Language.strScreenshots
                Text = Language.Language.strScreenshots
            End Sub

#End Region

#Region "Public Methods"

            Public Sub New(Panel As DockContent)
                Me.WindowType = Type.ScreenshotManager
                Me.DockPnl = Panel
                Me.InitializeComponent()
            End Sub

            Public Sub AddScreenshot(Screenshot As Image)
                Try
                    Dim nPB As New PictureBox
                    With nPB
                        AddHandler .MouseDown, AddressOf Me.pbScreenshot_MouseDown

                        .Parent = Me.flpScreenshots
                        .SizeMode = PictureBoxSizeMode.StretchImage
                        .BorderStyle = BorderStyle.FixedSingle
                        .ContextMenuStrip = Me.cMenScreenshot
                        .Image = Screenshot
                        .Size = New Size(100, 100) _
                        'New Size((Screenshot.Width / 100) * 20, (Screenshot.Height / 100) * 20)
                        .Show()
                    End With

                    Dim nBtn As New Button
                    With nBtn
                        AddHandler .Click, AddressOf btnCloseScreenshot_Click

                        .Parent = nPB
                        .FlatStyle = FlatStyle.Flat
                        .Text = "×"
                        .Size = New Size(22, 22)
                        .Location = New Point(nPB.Width - .Width, - 1)
                        .Show()
                    End With

                    Me.Show(frmMain.pnlDock)
                Catch ex As Exception
                   App.Runtime.MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "AddScreenshot (UI.Window.ScreenshotManager) failed" & vbNewLine & ex.ToString(), True)
                End Try
            End Sub

#End Region

#Region "Private Methods"

            Private Sub pbScreenshot_MouseDown(sender As Object, e As MouseEventArgs)
                Me.cMenScreenshot.Tag = sender

                If e.Button = MouseButtons.Left Then
                    Me.OpenScreenshot(sender)
                End If
            End Sub

            Private Sub pbScreenshotOpen_MouseDown(sender As Object, e As MouseEventArgs)
                If e.Button = MouseButtons.Left Then
                    Me.CloseOpenedScreenshot(sender.Parent)
                End If
            End Sub

            Private Sub CloseOpenedScreenshot(form As Form)
                form.Close()
            End Sub

            Private Sub OpenScreenshot(sender As PictureBox)
                Try
                    Dim mImage As Image = sender.Image

                    Dim nForm As New Form()
                    nForm.StartPosition = FormStartPosition.CenterParent
                    nForm.ShowInTaskbar = False
                    nForm.ShowIcon = False
                    nForm.MaximizeBox = False
                    nForm.MinimizeBox = False
                    nForm.Width = mImage.Width + 2
                    nForm.Height = mImage.Height + 2
                    nForm.FormBorderStyle = FormBorderStyle.None

                    Dim nPB As New PictureBox
                    nPB.Parent = nForm
                    nPB.BorderStyle = BorderStyle.FixedSingle
                    nPB.Location = New Point(0, 0)
                    nPB.SizeMode = PictureBoxSizeMode.AutoSize
                    nPB.Anchor = AnchorStyles.Left Or AnchorStyles.Bottom Or AnchorStyles.Right Or AnchorStyles.Top
                    nPB.Image = mImage
                    nPB.ContextMenuStrip = Me.cMenScreenshot
                    nPB.Show()

                    AddHandler nPB.MouseDown, AddressOf Me.pbScreenshotOpen_MouseDown

                    nForm.ShowDialog()
                Catch ex As Exception
                   App.Runtime.MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "OpenScreenshot (UI.Window.ScreenshotManager) failed" & vbNewLine & ex.ToString(), True)
                End Try
            End Sub

            Private Sub btnCloseScreenshot_Click(sender As Object, e As EventArgs)
                Try
                    sender.Parent.Dispose()
                Catch ex As Exception
                   App.Runtime.MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "btnCloseScreenshot_Click (UI.Window.ScreenshotManager) failed" & vbNewLine & ex.ToString(), True)
                End Try
            End Sub

            Private Sub mMenFileRemoveAll_Click(sender As Object, e As EventArgs) Handles mMenFileRemoveAll.Click
                Me.RemoveAllImages()
            End Sub

            Private Sub RemoveAllImages()
                Me.flpScreenshots.Controls.Clear()
            End Sub

            Private Sub mMenFileSaveAll_Click(sender As Object, e As EventArgs) Handles mMenFileSaveAll.Click
                Me.SaveAllImages()
            End Sub

            Private Sub SaveAllImages()
                Try
                    Dim pCount = 1

                    If Me.dlgSaveAllImages.ShowDialog = DialogResult.OK Then
                        For Each fPath As String In _
                            Directory.GetFiles(Me.dlgSaveAllImages.SelectedPath, "Screenshot_*",
                                               SearchOption.TopDirectoryOnly)
                            Dim f As New FileInfo(fPath)

                            Dim fCount As String = f.Name
                            fCount = fCount.Replace(f.Extension, "")
                            fCount = fCount.Replace("Screenshot_", "")

                            pCount = fCount + 1
                        Next

                        For Each ctrl As Control In Me.flpScreenshots.Controls
                            If TypeOf ctrl Is PictureBox Then
                                TryCast(ctrl, PictureBox).Image.Save(
                                    Me.dlgSaveAllImages.SelectedPath & "\Screenshot_" & Misc.LeadingZero(pCount) &
                                    ".png", ImageFormat.Png)
                                pCount += 1
                            End If
                        Next
                    End If
                Catch ex As Exception
                    Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                        "SaveAllImages (UI.Window.ScreenshotManager) failed" & vbNewLine &
                                                        ex.ToString(), True)
                End Try
            End Sub

            Private Sub cMenScreenshotCopy_Click(sender As Object, e As EventArgs) Handles cMenScreenshotCopy.Click
                Me.CopyImageToClipboard()
            End Sub

            Private Sub CopyImageToClipboard()
                Try
                    Clipboard.SetImage(TryCast(cMenScreenshot.Tag, PictureBox).Image)
                Catch ex As Exception
                   App.Runtime.MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "CopyImageToClipboard (UI.Window.ScreenshotManager) failed" & vbNewLine & ex.ToString(), True)
                End Try
            End Sub

            Private Sub cMenScreenshotSave_Click(sender As Object, e As EventArgs) Handles cMenScreenshotSave.Click
                Me.SaveSingleImage()
            End Sub

            Private Sub SaveSingleImage()
                Try
                    If Me.dlgSaveSingleImage.ShowDialog() = DialogResult.OK Then
                        Select Case _
                            LCase(
                                Me.dlgSaveSingleImage.FileName.Substring(
                                    Me.dlgSaveSingleImage.FileName.LastIndexOf(".") + 1))
                            Case "gif"
                                TryCast(Me.cMenScreenshot.Tag, PictureBox).Image.Save(Me.dlgSaveSingleImage.FileName,
                                                                                      ImageFormat.Gif)
                            Case "jpeg"
                                TryCast(Me.cMenScreenshot.Tag, PictureBox).Image.Save(Me.dlgSaveSingleImage.FileName,
                                                                                      ImageFormat.Jpeg)
                            Case "jpg"
                                TryCast(Me.cMenScreenshot.Tag, PictureBox).Image.Save(Me.dlgSaveSingleImage.FileName,
                                                                                      ImageFormat.Jpeg)
                            Case "png"
                                TryCast(Me.cMenScreenshot.Tag, PictureBox).Image.Save(Me.dlgSaveSingleImage.FileName,
                                                                                      ImageFormat.Png)
                        End Select
                    End If
                Catch ex As Exception
                    Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                        "SaveSingleImage (UI.Window.ScreenshotManager) failed" &
                                                        vbNewLine & ex.ToString(), True)
                End Try
            End Sub

            Private Sub mMenFile_DropDownOpening(sender As Object, e As EventArgs) Handles mMenFile.DropDownOpening
                If Me.flpScreenshots.Controls.Count < 1 Then
                    Me.mMenFileSaveAll.Enabled = False
                    Me.mMenFileRemoveAll.Enabled = False
                Else
                    Me.mMenFileSaveAll.Enabled = True
                    Me.mMenFileRemoveAll.Enabled = True
                End If
            End Sub

#End Region
        End Class
    End Namespace

End Namespace

File: mRemoteV1.ConvertedToC#/UI/UI.Window.ScreenshotManager.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using WeifenLuo.WinFormsUI.Docking;
using mRemoteNG.App.Runtime;
using System.IO;

namespace mRemoteNG.UI
{
	namespace Window
	{
		public class ScreenshotManager : UI.Window.Base
		{

			#region "Form Init"
			internal System.Windows.Forms.MenuStrip msMain;
			private System.Windows.Forms.ToolStripMenuItem withEventsField_mMenFile;
			internal System.Windows.Forms.ToolStripMenuItem mMenFile {
				get { return withEventsField_mMenFile; }
				set {
					if (withEventsField_mMenFile != null) {
						withEventsField_mMenFile.DropDownOpening -= mMenFile_DropDownOpening;
					}
					withEventsField_mMenFile = value;
					if (withEventsField_mMenFile != null) {
						withEventsField_mMenFile.DropDownOpening += mMenFile_DropDownOpening;
					}
				}
			}
			private System.Windows.Forms.ToolStripMenuItem withEventsField_mMenFileSaveAll;
			internal System.Windows.Forms.ToolStripMenuItem mMenFileSaveAll {
				get { return withEventsField_mMenFileSaveAll; }
				set {
					if (withEventsField_mMenFileSaveAll != null) {
						withEventsField_mMenFileSaveAll.Click -= mMenFileSaveAll_Click;
					}
					withEventsField_mMenFileSaveAll = value;
					if (withEventsField_mMenFileSaveAll != null) {
						withEventsField_mMenFileSaveAll.Click += mMenFileSaveAll_Click;
					}
				}
			}
			private System.Windows.Forms.ToolStripMenuItem withEventsField_mMenFileRemoveAll;
			internal System.Windows.Forms.ToolStripMenuItem mMenFileRemoveAll {
				get { return withEventsField_mMenFileRemoveAll; }
				set {
					if (withEventsField_mMenFileRemoveAll != null) {
						withEventsField_mMenFileRemoveAll.Click -= mMenFileRemoveAll_Click;
					}
					withEventsField_mMenFileRemoveAll = value;
					if (withEventsField_mMenFileRemoveAll != null) {
						withEventsField_mMenFileRemoveAll.Click += mMenFileRemoveAll_Click;
					}
				}
			}
			internal System.Windows.Forms.ContextMenuStrip cMenScreenshot;
			private System.ComponentModel.IContainer components;
			private System.Windows.Forms.ToolStripMenuItem withEventsField_cMenScreenshotCopy;
			internal System.Windows.Forms.ToolStripMenuItem cMenScreenshotCopy {
				get { return withEventsField_cMenScreenshotCopy; }
				set {
					if (withEventsField_cMenScreenshotCopy != null) {
						withEventsField_cMenScreenshotCopy.Click -= cMenScreenshotCopy_Click;
					}
					withEventsField_cMenScreenshotCopy = value;
					if (withEventsField_cMenScreenshotCopy != null) {
						withEventsField_cMenScreenshotCopy.Click += cMenScreenshotCopy_Click;
					}
				}
			}
			private System.Windows.Forms.ToolStripMenuItem withEventsField_cMenScreenshotSave;
			internal System.Windows.Forms.ToolStripMenuItem cMenScreenshotSave {
				get { return withEventsField_cMenScreenshotSave; }
				set {
					if (withEventsField_cMenScreenshotSave != null) {
						withEventsField_cMenScreenshotSave.Click -= cMenScreenshotSave_Click;
					}
					withEventsField_cMenScreenshotSave = value;
					if (withEventsField_cMenScreenshotSave != null) {
						withEventsField_cMenScreenshotSave.Click += cMenScreenshotSave_Click;
					}
				}
			}
			internal System.Windows.Forms.SaveFileDialog dlgSaveSingleImage;
			internal System.Windows.Forms.FolderBrowserDialog dlgSaveAllImages;

			internal System.Windows.Forms.FlowLayoutPanel flpScreenshots;
			private void InitializeComponent()
			{
				this.components = new System.ComponentModel.Container();
				this.flpScreenshots = new System.Windows.Forms.FlowLayoutPanel();
				this.msMain = new System.Windows.Forms.MenuStrip();
				this.mMenFile = new System.Windows.Forms.ToolStripMenuItem();
				this.mMenFileSaveAll = new System.Windows.Forms.ToolStripMenuItem();
				this.mMenFileRemoveAll = new System.Windows.Forms.ToolStripMenuItem();
				this.cMenScreenshot = new System.Windows.Forms.ContextMenuStrip(this.components);
				this.cMenScreenshotCopy = new System.Windows.Forms.ToolStripMenuItem();
				this.cMenScreenshotSave = new System.Windows.Forms.ToolStripMenuItem();
				this.dlgSaveSingleImage = new System.Windows.Forms.SaveFileDialog();
				this.dlgSaveAllImages = new System.Windows.Forms.FolderBrowserDialog();
				this.msMain.SuspendLayout();
				this.cMenScreenshot.SuspendLayout();
				this.SuspendLayout();
				//
				//flpScreenshots
				//
				this.flpScreenshots.Anchor = (System.Windows.Forms.AnchorStyles)(((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Bottom) | System.Windows.Forms.AnchorStyles.Left) | System.Windows.Forms.AnchorStyles.Right);
				this.flpScreenshots.AutoScroll = true;
				this.flpScreenshots.Location = new System.Drawing.Point(0, 26);
				this.flpScreenshots.Name = "flpScreenshots";
				this.flpScreenshots.Size = new System.Drawing.Size(542, 296);
				this.flpScreenshots.TabIndex = 0;
				//
				//msMain
				//
				this.msMain.Font = new System.Drawing.Font("Microsoft Sans Serif", 8.25f, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, Convert.ToByte(0));
				this.msMain.Items.AddRange(new System.Windows.Forms.ToolStripItem[] { this.mMenFile });
				this.msMain.Location = new System.Drawing.Point(0, 0);
				this.msMain.Name = "msMain";
				this.msMain.RenderMode = System.Windows.Forms.ToolStripRenderMode.Professional;
				this.msMain.Size = new System.Drawing.Size(542, 24);
				this.msMain.TabIndex = 1;
				this.msMain.Text = "MenuStrip1";
				//
				//mMenFile
				//
				this.mMenFile.DropDownItems.AddRange(new System.Windows.Forms.ToolStripItem[] {
					this.mMenFileSaveAll,
					this.mMenFileRemoveAll
				});
				this.mMenFile.Image = global::mRemoteNG.My.Resources.Resources.File;
				this.mMenFile.Name = "mMenFile";
				this.mMenFile.Size = new System.Drawing.Size(51, 20);
				this.mMenFile.Text = "&File";
				//
				//mMenFileSaveAll
				//
				this.mMenFileSaveAll.Image = global::mRemoteNG.My.Resources.Resources.Screenshot_Save;
				this.mMenFileSaveAll.Name = "mMenFileSaveAll";
				this.mMenFileSaveAll.Size = new System.Drawing.Size(128, 22);
				this.mMenFileSaveAll.Text = "Save All";
				//
				//mMenFileRemoveAll
				//
				this.mMenFileRemoveAll.Image = global::mRemoteNG.My.Resources.Resources.Screenshot_Delete;
				this.mMenFileRemoveAll.Name = "mMenFileRemoveAll";
				this.mMenFileRemoveAll.Size = new System.Drawing.Size(128, 22);
				this.mMenFileRemoveAll.Text = "Remove All";
				//
				//cMenScreenshot
				//
				this.cMenScreenshot.Items.AddRange(new System.Windows.Forms.ToolStripItem[] {
					this.cMenScreenshotCopy,
					this.cMenScreenshotSave
				});
				this.cMenScreenshot.Name = "cMenScreenshot";
				this.cMenScreenshot.Size = new System.Drawing.Size(100, 48);
				//
				//cMenScreenshotCopy
				//
				this.cMenScreenshotCopy.Image = global::mRemoteNG.My.Resources.Resources.Screenshot_Copy;
				this.cMenScreenshotCopy.Name = "cMenScreenshotCopy";
				this.cMenScreenshotCopy.Size = new System.Drawing.Size(99, 22);
				this.cMenScreenshotCopy.Text = "Copy";
				//
				//cMenScreenshotSave
				//
				this.cMenScreenshotSave.Image = global::mRemoteNG.My.Resources.Resources.Screenshot_Save;
				this.cMenScreenshotSave.Name = "cMenScreenshotSave";
				this.cMenScreenshotSave.Size = new System.Drawing.Size(99, 22);
				this.cMenScreenshotSave.Text = "Save";
				//
				//dlgSaveSingleImage
				//
				this.dlgSaveSingleImage.Filter = "Graphics Interchange Format File (.gif)|*.gif|Joint Photographic Experts Group Fi" + "le (.jpeg)|*.jpeg|Joint Photographic Experts Group File (.jpg)|*.jpg|Portable Ne" + "twork Graphics File (.png)|*.png";
				this.dlgSaveSingleImage.FilterIndex = 4;
				//
				//ScreenshotManager
				//
				this.ClientSize = new System.Drawing.Size(542, 323);
				this.Controls.Add(this.flpScreenshots);
				this.Controls.Add(this.msMain);
				this.HideOnClose = true;
				this.Icon = global::mRemoteNG.My.Resources.Resources.Screenshot_Icon;
				this.MainMenuStrip = this.msMain;
				this.Name = "ScreenshotManager";
				this.TabText = "Screenshots";
				this.Text = "Screenshots";
				this.msMain.ResumeLayout(false);
				this.msMain.PerformLayout();
				this.cMenScreenshot.ResumeLayout(false);
				this.ResumeLayout(false);
				this.PerformLayout();

			}
			#endregion

			#region "Form Stuff"
			private void ScreenshotManager_Load(object sender, System.EventArgs e)
			{
				ApplyLanguage();
			}

			private void ApplyLanguage()
			{
				mMenFile.Text = mRemoteNG.My.Language.strMenuFile;
				mMenFileSaveAll.Text = mRemoteNG.My.Language.strSaveAll;
				mMenFileRemoveAll.Text = mRemoteNG.My.Language.strRemoveAll;
				cMenScreenshotCopy.Text = mRemoteNG.My.Language.strMenuCopy;
				cMenScreenshotSave.Text = mRemoteNG.My.Language.strSave;
				dlgSaveSingleImage.Filter = mRemoteNG.My.Language.strSaveImageFilter;
				TabText = mRemoteNG.My.Language.strScreenshots;
				Text = mRemoteNG.My.Language.strScreenshots;
			}
			#endregion

			#region "Public Methods"
			public ScreenshotManager(DockContent Panel)
			{
				Load += ScreenshotManager_Load;
				this.WindowType = Type.ScreenshotManager;
				this.DockPnl = Panel;
				this.InitializeComponent();
			}

			public void AddScreenshot(Image Screenshot)
			{
				try {
					PictureBox nPB = new PictureBox();
					var _with1 = nPB;
					_with1.MouseDown += this.pbScreenshot_MouseDown;

					_with1.Parent = this.flpScreenshots;
					_with1.SizeMode = PictureBoxSizeMode.StretchImage;
					_with1.BorderStyle = BorderStyle.FixedSingle;
					_with1.ContextMenuStrip = this.cMenScreenshot;
					_with1.Image = Screenshot;
					_with1.Size = new Size(100, 100);
					//New Size((Screenshot.Width / 100) * 20, (Screenshot.Height / 100) * 20)
					_with1.Show();

					Button nBtn = new Button();
					var _with2 = nBtn;
					_with2.Click += btnCloseScreenshot_Click;

					_with2.Parent = nPB;
					_with2.FlatStyle = FlatStyle.Flat;
					_with2.Text = "×";
					_with2.Size = new Size(22, 22);
					_with2.Location = new Point(nPB.Width - _with2.Width, -1);
					_with2.Show();

					this.Show(My.MyProject.Forms.frmMain.pnlDock);
				} catch (Exception ex) {
					mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.ErrorMsg, "AddScreenshot (UI.Window.ScreenshotManager) failed" + Constants.vbNewLine + ex.Message, true);
				}
			}
			#endregion

			#region "Private Methods"
			private void pbScreenshot_MouseDown(System.Object sender, System.Windows.Forms.MouseEventArgs e)
			{
				this.cMenScreenshot.Tag = sender;

				if (e.Button == System.Windows.Forms.MouseButtons.Left) {
					this.OpenScreenshot(sender);
				}
			}

			private void pbScreenshotOpen_MouseDown(System.Object sender, System.Windows.Forms.MouseEventArgs e)
			{
				if (e.Button == System.Windows.Forms.MouseButtons.Left) {
					this.CloseOpenedScreenshot(sender.Parent);
				}
			}

			private void CloseOpenedScreenshot(Form form)
			{
				form.Close();
			}

			private void OpenScreenshot(PictureBox sender)
			{
				try {
					Image mImage = sender.Image;

					Form nForm = new Form();
					nForm.StartPosition = FormStartPosition.CenterParent;
					nForm.ShowInTaskbar = false;
					nForm.ShowIcon = false;
					nForm.MaximizeBox = false;
					nForm.MinimizeBox = false;
					nForm.Width = mImage.Width + 2;
					nForm.Height = mImage.Height + 2;
					nForm.FormBorderStyle = System.Windows.Forms.FormBorderStyle.None;

					PictureBox nPB = new PictureBox();
					nPB.Parent = nForm;
					nPB.BorderStyle = BorderStyle.FixedSingle;
					nPB.Location = new Point(0, 0);
					nPB.SizeMode = PictureBoxSizeMode.AutoSize;
					nPB.Anchor = AnchorStyles.Left | AnchorStyles.Bottom | AnchorStyles.Right | AnchorStyles.Top;
					nPB.Image = mImage;
					nPB.ContextMenuStrip = this.cMenScreenshot;
					nPB.Show();

					nPB.MouseDown += this.pbScreenshotOpen_MouseDown;

					nForm.ShowDialog();
				} catch (Exception ex) {
					mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.ErrorMsg, "OpenScreenshot (UI.Window.ScreenshotManager) failed" + Constants.vbNewLine + ex.Message, true);
				}
			}

			private void btnCloseScreenshot_Click(System.Object sender, System.EventArgs e)
			{
				try {
					sender.Parent.Dispose();
				} catch (Exception ex) {
					mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.ErrorMsg, "btnCloseScreenshot_Click (UI.Window.ScreenshotManager) failed" + Constants.vbNewLine + ex.Message, true);
				}
			}

			private void mMenFileRemoveAll_Click(System.Object sender, System.EventArgs e)
			{
				this.RemoveAllImages();
			}

			private void RemoveAllImages()
			{
				this.flpScreenshots.Controls.Clear();
			}

			private void mMenFileSaveAll_Click(System.Object sender, System.EventArgs e)
			{
				this.SaveAllImages();
			}

			private void SaveAllImages()
			{
				try {
					int pCount = 1;

					if (this.dlgSaveAllImages.ShowDialog() == System.Windows.Forms.DialogResult.OK) {
						foreach (string fPath in Directory.GetFiles(this.dlgSaveAllImages.SelectedPath, "Screenshot_*", SearchOption.TopDirectoryOnly)) {
							FileInfo f = new FileInfo(fPath);

							string fCount = f.Name;
							fCount = fCount.Replace(f.Extension, "");
							fCount = fCount.Replace("Screenshot_", "");

							pCount = fCount + 1;
						}

						foreach (System.Windows.Forms.Control ctrl in this.flpScreenshots.Controls) {
							if (ctrl is PictureBox) {
								(ctrl as PictureBox).Image.Save(this.dlgSaveAllImages.SelectedPath + "\\Screenshot_" + mRemoteNG.Tools.Misc.LeadingZero(pCount) + ".png", System.Drawing.Imaging.ImageFormat.Png);
								pCount += 1;
							}
						}
					}
				} catch (Exception ex) {
					mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.ErrorMsg, "SaveAllImages (UI.Window.ScreenshotManager) failed" + Constants.vbNewLine + ex.Message, true);
				}
			}

			private void cMenScreenshotCopy_Click(System.Object sender, System.EventArgs e)
			{
				this.CopyImageToClipboard();
			}

			private void CopyImageToClipboard()
			{
				try {
					Clipboard.SetImage((cMenScreenshot.Tag as PictureBox).Image);
				} catch (Exception ex) {
					mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.ErrorMsg, "CopyImageToClipboard (UI.Window.ScreenshotManager) failed" + Constants.vbNewLine + ex.Message, true);
				}
			}

			private void cMenScreenshotSave_Click(System.Object sender, System.EventArgs e)
			{
				this.SaveSingleImage();
			}

			private void SaveSingleImage()
			{
				try {
					if (this.dlgSaveSingleImage.ShowDialog() == System.Windows.Forms.DialogResult.OK) {
						switch (Strings.LCase(this.dlgSaveSingleImage.FileName.Substring(this.dlgSaveSingleImage.FileName.LastIndexOf(".") + 1))) {
							case "gif":
								(this.cMenScreenshot.Tag as PictureBox).Image.Save(this.dlgSaveSingleImage.FileName, System.Drawing.Imaging.ImageFormat.Gif);
								break;
							case "jpeg":
								(this.cMenScreenshot.Tag as PictureBox).Image.Save(this.dlgSaveSingleImage.FileName, System.Drawing.Imaging.ImageFormat.Jpeg);
								break;
							case "jpg":
								(this.cMenScreenshot.Tag as PictureBox).Image.Save(this.dlgSaveSingleImage.FileName, System.Drawing.Imaging.ImageFormat.Jpeg);
								break;
							case "png":
								(this.cMenScreenshot.Tag as PictureBox).Image.Save(this.dlgSaveSingleImage.FileName, System.Drawing.Imaging.ImageFormat.Png);
								break;
						}
					}
				} catch (Exception ex) {
					mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.ErrorMsg, "SaveSingleImage (UI.Window.ScreenshotManager) failed" + Constants.vbNewLine + ex.Message, true);
				}
			}

			private void mMenFile_DropDownOpening(object sender, System.EventArgs e)
			{
				if (this.flpScreenshots.Controls.Count < 1) {
					this.mMenFileSaveAll.Enabled = false;
					this.mMenFileRemoveAll.Enabled = false;
				} else {
					this.mMenFileSaveAll.Enabled = true;
					this.mMenFileRemoveAll.Enabled = true;
				}
			}
			#endregion

		}
	}
}
